commit d94c6a1a6414ec735ada7fb13e01a19d5094beab (refs/changes/73/1173/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-22T15:37:26-08:00 (2 years, 10 months ago)
    
    joyent/node-mname#12 label runs over end of packet, readName() runs off the rails

diff --git a/lib/dns-buffer.js b/lib/dns-buffer.js
index 5e45df4..5d3a3bc 100644
--- a/lib/dns-buffer.js
+++ b/lib/dns-buffer.js
@@ -78,7 +78,7 @@ DNSBuffer.prototype.readUInt16 = function () {
 };
 
 DNSBuffer.prototype.readUInt8 = function () {
-        var v = this._buffer[this._offset++];
+        var v = this._buffer.readUInt8(this._offset++);
         return (v);
 };
 
@@ -94,6 +94,8 @@ DNSBuffer.prototype.readName = function () {
                 var meta = rlen & NAME_META_MASK;
 
                 if (meta == NAME_STRING) {
+                        assert.ok(this._offset + rlen < this._size,
+                            'invalid name label length');
                         var buf = this._buffer.slice(this._offset,
                             this._offset + rlen);
                         this._offset += rlen;
@@ -103,6 +105,9 @@ DNSBuffer.prototype.readName = function () {
                         var ptr = this.readUInt8();
                         ptr = ptr | ((rlen & ~(0xC0)) << 8);
 
+                        assert.ok(ptr < this._size,
+                            'invalid label pointer (off end of buf)');
+
                         var clone = Object.create(this);
                         clone._offset = ptr;
                         name.push(clone.readName());
diff --git a/package.json b/package.json
index 0025677..374a491 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "mname",
   "description": "DNS server library for node.js",
-  "version": "1.3.1",
+  "version": "1.3.2",
   "author": "arekinath <alex@cooperi.net>",
   "contributors": [
     "Mark Cavage",
diff --git a/test/protocol.test.js b/test/protocol.test.js
index f670811..935b356 100644
--- a/test/protocol.test.js
+++ b/test/protocol.test.js
@@ -35,7 +35,7 @@ var protocol = named.Protocol;
 Object.keys(dnsBuffer.samples).forEach(function (i) {
         var sample = dnsBuffer.samples[i];
         test('protocol decode/encode: ' + sample.description, function(t) {
-                decoded = protocol.decode(sample.raw, sample.type);
+                var decoded = protocol.decode(sample.raw, sample.type);
                 if (sample.decoded !== undefined) {
                         t.deepEqual(decoded, sample.decoded);
                 }
@@ -56,3 +56,11 @@ Object.keys(dnsBuffer.samples).forEach(function (i) {
                 t.end();
         });
 });
+
+test('mname#12 regression test', function (t) {
+        var b = new Buffer('GET / HTTP/1.1\r\n\r\n');
+        t.throws(function () {
+                var decoded = protocol.decode(b, 'message');
+        });
+        t.end();
+});
